mm2src/coins/utxo/tx_cache.rs

```diff
@@ -1,42 +1,82 @@
-use common::fs::safe_slurp;
+use common::fs::{read_json, write_json, FsJsonError};
+use common::log::LogOnError;
+use common::mm_error::prelude::*;
+use derive_more::Display;
 use futures::lock::Mutex as AsyncMutex;
+use futures::FutureExt;
 use rpc::v1::types::{Transaction as RpcTransaction, H256 as H256Json};
+use std::collections::HashMap;
 use std::path::{Path, PathBuf};
 
+pub type TxCacheResult<T> = MmResult<T, TxCacheError>;
+
 lazy_static! {
     static ref TX_CACHE_LOCK: AsyncMutex<()> = AsyncMutex::new(());
 }
 
-/// Try load transaction from cache.
-/// Note: tx.confirmations can be out-of-date.
-pub async fn load_transaction_from_cache(
-    tx_cache_path: &Path,
-    txid: &H256Json,
-) -> Result<Option<RpcTransaction>, String> {
-    let _lock = TX_CACHE_LOCK.lock().await;
-
-    let path = cached_transaction_path(tx_cache_path, txid);
-    let data = try_s!(safe_slurp(&path));
-    if data.is_empty() {
-        // couldn't find corresponding file
-        return Ok(None);
+#[derive(Debug, Display)]
+pub enum TxCacheError {
+    ErrorLoading(String),
+    ErrorSaving(String),
+    ErrorDeserializing(String),
+    ErrorSerializing(String),
+}
+
+impl From<FsJsonError> for TxCacheError {
+    fn from(e: FsJsonError) -> Self {
+        match e {
+            FsJsonError::IoReading(loading) => TxCacheError::ErrorLoading(loading.to_string()),
+            FsJsonError::IoWriting(writing) => TxCacheError::ErrorSaving(writing.to_string()),
+            FsJsonError::Serializing(ser) => TxCacheError::ErrorSerializing(ser.to_string()),
+            FsJsonError::Deserializing(de) => TxCacheError::ErrorDeserializing(de.to_string()),
+        }
     }
+}
+
+/// Tries to load transactions from cache concurrently.
+/// Note 1: tx.confirmations can be out-of-date.
+/// Note 2: this function locks the `TX_CACHE_LOCK` mutex to avoid reading and writing the same files at the same time.
+pub async fn load_transactions_from_cache_concurrently<I>(
+    tx_cache_path: &Path,
+    tx_ids: I,
+) -> HashMap<H256Json, TxCacheResult<Option<RpcTransaction>>>
+where
+    I: IntoIterator<Item = H256Json>,
+{
+    let _ = TX_CACHE_LOCK.lock().await;
```

Comment:
The global lock for all coins will possibly become a bottleneck. It should be unique per each `tx_cache_path` or so.

Reply:
TX cache is designed so that `QTUM` cached transactions are used by `QRC20` tokens and vice versa.
I agree that it would be great to avoid locking the global mutex.
What do you think about adding a structure like
```rust
lazy_static! {
    static ref TX_CACHE_LOCK: TxCacheLock = TxCacheLock::default();
}

#[derive(Default)]
struct TxCacheLock {
    mutexes: PaMutex<HashMap<String, Arc<AsyncMutex<()>>>>,
}

impl TxCacheLock {
    pub fn mutex_by_ticker(&self, ticker: String) -> Arc<AsyncMutex<()>> {
        let mut locks = self.mutexes.lock();
        locks
            .entry(ticker)
            .or_insert_with(|| Arc::new(AsyncMutex::new(())))
            .clone()
    }
}

pub struct UtxoVerboseCache {
    ticker: String,
    tx_cache_path: PathBuf,
}

impl UtxoVerboseCache {
    pub async fn load_transactions_from_cache_concurrently<I>(
        &self,
        tx_ids: I,
    ) -> HashMap<H256Json, TxCacheResult<Option<RpcTransaction>>>
    where
        I: IntoIterator<Item = H256Json>,
    {
        let mutex = TX_CACHE_LOCK.mutex_by_ticker(self.ticker.clone());
        let _lock = mutex.lock().await;

        let it = tx_ids
            .into_iter()
            .map(|txid| load_transaction_from_cache(&self.tx_cache_path, txid).map(move |res| (txid, res)));
        futures::future::join_all(it).await.into_iter().collect()
    }
}
```
In this case `TX_CACHE_LOCK` is a global mutex that is locked for a short time. For `QRC20` token we can use [Qrc20CoinFields::platform](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/qrc20.rs#L293) as a ticker.

Reply:
> TX cache is designed so that QTUM cached transactions are used by QRC20 tokens and vice versa.

Then this functionality should have been only a part of QTUM and QRC20 implementations. But `get_verbose_transactions_from_cache_or_rpc` is part of `UtxoCommonOps`, which is implemented for many coin protocols. I have used it already for `BchCoin` and we don't know for how many coins we will use it too in the future as nothing prevents it.

> In this case TX_CACHE_LOCK is a global mutex that is locked for a short time. For QRC20 token we can use [Qrc20CoinFields::platform](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/coins/qrc20.rs#L293) as a ticker.

Should be ok for now. We should also consider a more general QRC20 and ERC20 refactoring, similarly to how it's done for SLP.

Reply:
>  But get_verbose_transactions_from_cache_or_rpc is part of UtxoCommonOps, which is implemented for many coin protocols.

It's even used by any UTXO coin if `check_utxo_maturity: true`

> We should also consider a more general QRC20 and ERC20 refactoring, similarly to how it's done for SLP.

I totally agree.

Reply:
Done